firebase_admin/dynamic_links.py

```diff
@@ -0,0 +1,215 @@
+            raise ValueError('Invalid data argument: {0}. Must be a list or tuple'
+                             .format(event_stats))
+        if event_stats and not isinstance(event_stats[0], EventStats):
+            raise ValueError('Invalid data argument: elements of event stats must be' +
+                             ' "EventStats", found{}'.format(type(event_stats[0])))
+        self._stats = event_stats
+
+    @property
+    def event_stats(self):
+        """Returns the event_stats for this link.
+
+        Returns:
+          event_stats: A list of EventStats.
+        """
+        return self._stats
+
+
+class _LinksService(object):
+    """Provides methods for the Firebase Dynamic Links interaction"""
+    def __init__(self, app):
+        self._client = _http_client.JsonHttpClient(
+            credential=app.credential.get_credential(),
+            base_url=_LINKS_BASE_URL)
+        self._timeout = app.options.get('httpTimeout')
+        self._links_request = '{0}/linkStats?durationDays={1}'
+
+    def _populated_request(self, url, options):
+        days = options.duration_days
+        # This is due to six.urllib_parse mistaken error for python 2
+        #pylint: disable=too-many-function-args
+        # Complaints about the named second argument needed to replace "/"
+        #pylint: disable=redundant-keyword-arg
+        url_quoted = urllib_parse.quote(url, safe="")
+        #pylint: enable=redundant-keyword-arg
+        #pylint: enable=too-many-function-args
+        return self._links_request.format(url_quoted, days)
+
+    def get_stats(self, url, options):
+        _validate_url(url)
+        _validate_stat_options(options)
+        url_p = self._populated_request(url, options)
+        resp = self._client.request('get', url_p)
+        link_event_stats = resp.json().get('linkEventStats', [])
+        event_stats = [EventStats.from_json(**es) for es in link_event_stats]
+
+        return LinkStats(event_stats)
+
+class EventStats(object):
+    """EventStats is a single stat item containing (platform, event, count)"""
+    _platforms = {
+        'DESKTOP': PLATFORM_DESKTOP,
+        'IOS': PLATFORM_IOS,
+        'ANDROID': PLATFORM_ANDROID
+    }
+
+    _event_types = {
+        'CLICK': EVENT_TYPE_CLICK,
+        'REDIRECT': EVENT_TYPE_REDIRECT,
+        'APP_INSTALL': EVENT_TYPE_APP_INSTALL,
+        'APP_FIRST_OPEN': EVENT_TYPE_APP_FIRST_OPEN,
+        'APP_RE_OPEN': EVENT_TYPE_APP_RE_OPEN
+    }
+
+    def __init__(self, platform, event, count):
+        """Create new instance of EventStats(platform, event, count)"""
+        self.platform = platform
+        self.event = event
+        self.count = count
+
+    def __repr__(self):
+        return"EventStats(platform: '{}', event: '{}', count: '{}')".format(
+            self.platform, self.event, self.count)
+
+    @classmethod
+    def from_json(cls, platform, event, count):
+        return EventStats(cls._platforms[platform],
```

Comment:
They are all required, I am not sure what we gain by this.

Reply:
If the server responds with an unexpected payload (which may not be a valid concern), this will fail with a TypeError or a KeyError. We can handle that case more gracefully, if this is implemented as suggested above.